README.md

```diff
@@ -150,9 +162,199 @@ class Foo {
+container and provides the container the information
+needed to resolve an instance for a given token. In our implementation, we have the following 4
+provider types:
+
+#### Class Provider
+
+```TypeScript
+{
+  token: InjectionToken<T>;
+  useClass: constructor<T>;
+}
+```
+
+This provider is used to resolve classes by their constructor. When registering a class provider
+you can simply use the constructor itself, unless of course you're making an alias (a
+class provider where the token isn't the class itself).
+
+#### Value Provider
+
+```TypeScript
+{
+  token: InjectionToken<T>;
+  useValue: T
+}
+```
+
+This provider is used to resolve a token to a given value. This is useful for registering
+constants, or things that have a already been instantiated in a particular way.
+
+#### Factory provider
+
+```TypeScript
+{
+  token: InjectionToken<T>;
+  useFactory: FactoryFunction<T>;
+}
+```
+
+This provider is used to resolve a token using a given factory. The factory has full access
+to the dependency container.
+
+We have provided 2 factories for you to use, though any function that matches the `FactoryFunction<T>` signature
+can be used as a factory:
+```typescript
+type FactoryFunction<T> = (dependencyContainer: DependencyContainer) => T;
+```
+
+##### instanceCachingFactory
+
+This factory is used to lazy construct an object and cache result, returning the single instance for each subsequent
+resolution. This is very similar to `@singleton()`
+
+```typescript
+import {instanceCachingFactory} from "tsyringe";
+
+{
+  token: "SingletonFoo";
+  useFactory: instanceCachingFactory<Foo>(c => c.resolve(Foo))
+}
+```
+
+##### predicateAwareClassFactory
+
+This factory is used to provide conditional behavior upon resolution. It caches the result by default, but 
+has an optional parameter to resolve fresh each time.
+
+```typescript
+import {predicateAwareClassFactory} from "tsyringe";
+
+{
+  token: 
+  useFactory: predicateAwareClassFactory<Foo>(
+    c => c.resolve(Bar).useHttps,
+    FooHttps, // A FooHttps will be resolved from the container
+    FooHttp
+  )
+}
+```
+#### Token Provider
+
+```TypeScript
+{
+  token: InjectionToken<T>;
+  useToken: InjectionToken<T>;
+}
+```
+
+This provider can be thought of as a redirect or an alias, it simply states that given token _x_,
+resolve using token _y_.
+
+### Register
+
+The normal way to achieve this is to add `DependencyContainer.register()` statements somewhere
+in your program some time before your first decorated class is instantiated.
+
+```typescript
+container.register<Foo>(Foo, {useClass: Foo});
+container.register<Bar>(Bar, {useValue: new Bar()});
+container.register<Baz>("MyBaz", {useValue: new Baz()});
+```
+
+### Registry
+
+You can also mark up any class with the `@registry()` decorator to have the given providers registered
+upon first use of the marked up class. `@registry()` takes an array of providers like so:
```

Comment:
technically I think its when the class is first imported, not when its first used. Minor but important difference

Reply:
I updated that line. It should reflect reality now.